modules/diskq/dqtool.c

```diff
@@ -241,12 +251,292 @@ _relocate_validate_options(void)
   return _validate_diskq_path(new_diskq_path) && _validate_persist_file_path(persist_file_path);
 }
 
+static gint
+_open_fd_for_reading(const gchar *fname)
+{
+  gint fd = open(fname, O_RDONLY);
+
+  if (fd == -1)
+    fprintf(stderr, "Failed to open %s, reason: %s\n", fname, strerror(errno));
+
+  return fd;
+}
+
+static gboolean
+_get_content_length_by_fd(gint fd, gsize *size)
+{
+  struct stat st;
+
+  if (fstat(fd, &st) == -1)
+    {
+      fprintf(stderr, "Error: %s\n", strerror(errno));
+      *size = 0;
+      return FALSE;
+    }
+
+  *size = st.st_size;
+  return TRUE;
+}
+
+static gboolean
+_mmap_content_from_fd(gint fd, gchar **content, gsize *content_length)
+{
+  if (!_get_content_length_by_fd(fd, content_length))
+    return FALSE;
+
+  *content = mmap(NULL, *content_length, PROT_READ, MAP_PRIVATE, fd, 0);
```

Comment:
Mapping a diskq file twice could be troublesome. Why did you choose to use `mmap + memcpy` instead of traditional `read + write` from a file ?